mergin/report.py

```diff
@@ -0,0 +1,237 @@
+import csv
+import json
+import os
+import tempfile
+from collections import defaultdict
+from itertools import groupby
+
+from . import ClientError
+from .merginproject import MerginProject, pygeodiff
+from .utils import int_version
+
+
+# inspired by C++ implementation https://github.com/lutraconsulting/geodiff/blob/master/geodiff/src/drivers/sqliteutils.cpp
+# in geodiff lib (MIT licence)
+# ideally there should be pygeodiff api for it
+def parse_gpkgb_header_size(gpkg_wkb):
+    """ Parse header of geopackage wkb and return its size """
+    # some constants
+    no_envelope_header_size = 8
+    flag_byte_pos = 3
+    envelope_size_mask = 14
+
+    try:
+        flag_byte = gpkg_wkb[flag_byte_pos]
+    except IndexError:
+        return -1  # probably some invalid input
+    envelope_byte = (flag_byte & envelope_size_mask) >> 1
+    envelope_size = 0
+
+    if envelope_byte == 1:
+        envelope_size = 32
+    elif envelope_byte == 2:
+        envelope_size = 48
+    elif envelope_byte == 3:
+        envelope_size = 48
+    elif envelope_byte == 4:
+        envelope_size = 64
+
+    return no_envelope_header_size + envelope_size
+
+
+class ChangesetReportEntry:
+    """ Derivative of geodiff ChangesetEntry suitable for further processing/reporting """
+    def __init__(self, changeset_entry, geom_idx, geom):
+        self.table = changeset_entry.table.name
+        self.geom_type = geom["type"]
+        self.crs = geom["srs_id"]
+
+        if changeset_entry.operation == changeset_entry.OP_DELETE:
+            self.operation = "delete"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = None
+        elif changeset_entry.operation == changeset_entry.OP_UPDATE:
+            self.operation = "update"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = changeset_entry.new_values[geom_idx]
+        elif changeset_entry.operation == changeset_entry.OP_INSERT:
+            self.operation = "insert"
+            self.old_geom = None
+            self.new_geom = changeset_entry.new_values[geom_idx]
+
+        self.count = None
+        self.length = None
+        self.area = None
+
+        if self.geom_type == "LINESTRING":
```

Comment:
there are more types to support: multi-part geometries (e.g. MULTILINESTRING) or geometries with extra dimensions (e.g. LINESTRINGZ, LINESTRINGZM, LINESTRINGM). Same for polygons...